File: frontend/src/services/procedureService.ts
import axios from 'axios';
import { API_BASE_URL } from '../config';

// Types
export enum ProcedureCategory {
  PREVENTIVE = 'preventive',
  RESTORATIVE = 'restorative',
  ENDODONTIC = 'endodontic',
  PERIODONTAL = 'periodontal',
  ORTHODONTIC = 'orthodontic',
  PROSTHODONTIC = 'prosthodontic',
  SURGICAL = 'surgical',
  OTHER = 'other'
}

export interface ProcedureBase {
  code: string;
  name: string;
  description: string;
  category: ProcedureCategory;
  default_duration_minutes: number;
}

export interface Procedure extends ProcedureBase {
  id: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProcedureRequirements {
  pre_operative: string[];
  post_operative: string[];
  equipment: string[];
  staff_requirements: number;
  room_requirements: string[];
}

export interface ProcedureStatistics {
  total_procedures: number;
  active_procedures: number;
  category_counts: Record<string, number>;
}

export interface ProviderDurationProfile {
  average_duration_minutes: number;
  duration_modifier: number;
  total_procedures: number;
}

class ProcedureService {
  private api = axios.create({
    baseURL: `${API_BASE_URL}/procedures`,
    headers: {
      'Content-Type': 'application/json',
    },
  });

  // List procedures with optional filtering
  async listProcedures(params?: {
    query?: string;
    category?: ProcedureCategory;
    active_only?: boolean;
  }): Promise<Procedure[]> {
    const response = await this.api.get<Procedure[]>('/', { params });
    return response.data;
  }

  // Get procedure by ID
  async getProcedure(id: string): Promise<Procedure> {
    const response = await this.api.get<Procedure>(`/${id}`);
    return response.data;
  }

  // Get procedure by ADA code
  async getProcedureByCode(code: string): Promise<Procedure> {
    const response = await this.api.get<Procedure>(`/code/${code}`);
    return response.data;
  }

  // Create new procedure
  async createProcedure(procedure: ProcedureBase): Promise<Procedure> {
    const response = await this.api.post<Procedure>('/', procedure);
    return response.data;
  }

  // Update existing procedure
  async updateProcedure(
    id: string,
    updates: Partial<ProcedureBase & { is_active?: boolean }>
  ): Promise<Procedure> {
    const response = await this.api.put<Procedure>(`/${id}`, updates);
    return response.data;
  }

  // Deactivate procedure
  async deactivateProcedure(id: string): Promise<void> {
    await this.api.delete(`/${id}`);
  }

  // Get procedure requirements
  async getProcedureRequirements(id: string): Promise<ProcedureRequirements> {
    const response = await this.api.get<ProcedureRequirements>(`/${id}/requirements`);
    return response.data;
  }

  // Get provider duration for a procedure
  async getProviderDuration(
    procedureId: string,
    providerId: string
  ): Promise<number> {
    const response = await this.api.get<{ duration_minutes: number }>(
      `/${procedureId}/duration/${providerId}`
    );
    return response.data.duration_minutes;
  }

  // Update provider duration tracking
  async updateProviderDuration(
    procedureId: string,
    providerId: string,
    actualDuration: number
  ): Promise<{
    message: string;
    profile: ProviderDurationProfile;
  }> {
    const response = await this.api.post<{
      message: string;
      profile: ProviderDurationProfile;
    }>(`/${procedureId}/duration/${providerId}`, null, {
      params: { actual_duration: actualDuration },
    });
    return response.data;
  }

  // Get procedure statistics
  async getStatistics(): Promise<ProcedureStatistics> {
    const response = await this.api.get<ProcedureStatistics>('/statistics');
    return response.data;
  }

  // Search procedures with debounce
  private searchTimeout: NodeJS.Timeout | null = null;
  async searchProcedures(
    query: string,
    category?: ProcedureCategory
  ): Promise<Procedure[]> {
    if (this.searchTimeout) {
      clearTimeout(this.searchTimeout);
    }

    return new Promise((resolve) => {
      this.searchTimeout = setTimeout(async () => {
        const procedures = await this.listProcedures({ query, category });
        resolve(procedures);
      }, 300); // 300ms debounce
    });
  }

  // Get procedures by category
  async getProceduresByCategory(category: ProcedureCategory): Promise<Procedure[]> {
    return this.listProcedures({ category, active_only: true });
  }

  // Get active procedures only
  async getActiveProcedures(): Promise<Procedure[]> {
    return this.listProcedures({ active_only: true });
  }

  // Get procedure duration with provider adjustment
  async getAdjustedDuration(
    procedureId: string,
    providerId: string
  ): Promise<{
    baseDuration: number;
    adjustedDuration: number;
    modifier: number;
  }> {
    const procedure = await this.getProcedure(procedureId);
    const providerDuration = await this.getProviderDuration(procedureId, providerId);
    
    return {
      baseDuration: procedure.default_duration_minutes,
      adjustedDuration: providerDuration,
      modifier: providerDuration / procedure.default_duration_minutes,
    };
  }
}

// Export singleton instance
export const procedureService = new ProcedureService(); 

File: backend/api/services/procedure_service.py
from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_
from api.models.scheduling import Procedure, ProviderProfile, ProcedureCategory
from api.models.base import Base
import logging
from datetime import datetime

logger = logging.getLogger(__name__)

class ProcedureService:
    def __init__(self, db: Session):
        self.db = db

    async def get_procedure_by_code(self, code: str) -> Optional[Procedure]:
        """Get a procedure by its ADA code."""
        return self.db.query(Procedure).filter(Procedure.code == code).first()

    async def get_procedure_by_id(self, procedure_id: str) -> Optional[Procedure]:
        """Get a procedure by its UUID."""
        return self.db.query(Procedure).filter(Procedure.id == procedure_id).first()

    async def search_procedures(
        self,
        query: Optional[str] = None,
        category: Optional[ProcedureCategory] = None,
        active_only: bool = True
    ) -> List[Procedure]:
        """Search procedures by name, code, or category."""
        filters = []
        if active_only:
            filters.append(Procedure.is_active == True)
        if query:
            filters.append(
                or_(
                    Procedure.name.ilike(f"%{query}%"),
                    Procedure.code.ilike(f"%{query}%"),
                    Procedure.description.ilike(f"%{query}%")
                )
            )
        if category:
            filters.append(Procedure.category == category)

        return self.db.query(Procedure).filter(and_(*filters)).all()

    async def get_provider_duration(
        self,
        procedure_id: str,
        provider_id: str
    ) -> int:
        """Calculate the expected duration for a provider performing a procedure."""
        # Get the base procedure
        procedure = await self.get_procedure_by_id(procedure_id)
        if not procedure:
            raise ValueError(f"Procedure {procedure_id} not found")

        # Get provider's profile for this procedure
        profile = self.db.query(ProviderProfile).filter(
            ProviderProfile.provider_id == provider_id,
            ProviderProfile.procedure_id == procedure_id,
            ProviderProfile.is_active == True
        ).first()

        if profile:
            # If we have enough data points, use the provider's average
            if profile.total_procedures >= 5 and profile.average_duration_minutes:
                return profile.average_duration_minutes
            # Otherwise, apply their modifier to the default duration
            return int(procedure.default_duration_minutes * profile.duration_modifier)
        
        # If no profile exists, return the default duration
        return procedure.default_duration_minutes

    async def update_provider_duration(
        self,
        procedure_id: str,
        provider_id: str,
        actual_duration: int
    ) -> ProviderProfile:
        """Update a provider's duration tracking for a procedure."""
        profile = self.db.query(ProviderProfile).filter(
            ProviderProfile.provider_id == provider_id,
            ProviderProfile.procedure_id == procedure_id
        ).first()

        if not profile:
            # Create new profile if none exists
            profile = ProviderProfile(
                provider_id=provider_id,
                procedure_id=procedure_id,
                total_procedures=1,
                average_duration_minutes=actual_duration,
                duration_modifier=1.0
            )
            self.db.add(profile)
        else:
            # Update existing profile
            total_duration = (profile.average_duration_minutes or 0) * profile.total_procedures
            profile.total_procedures += 1
            profile.average_duration_minutes = (total_duration + actual_duration) / profile.total_procedures
            profile.duration_modifier = profile.average_duration_minutes / profile.procedure.default_duration_minutes

        self.db.commit()
        return profile

    async def create_procedure(
        self,
        code: str,
        name: str,
        description: str,
        category: ProcedureCategory,
        default_duration_minutes: int
    ) -> Procedure:
        """Create a new procedure."""
        procedure = Procedure(
            code=code,
            name=name,
            description=description,
            category=category,
            default_duration_minutes=default_duration_minutes
        )
        self.db.add(procedure)
        self.db.commit()
        return procedure

    async def update_procedure(
        self,
        procedure_id: str,
        **kwargs
    ) -> Optional[Procedure]:
        """Update an existing procedure."""
        procedure = await self.get_procedure_by_id(procedure_id)
        if not procedure:
            return None

        for key, value in kwargs.items():
            if hasattr(procedure, key):
                setattr(procedure, key, value)

        procedure.updated_at = datetime.utcnow()
        self.db.commit()
        return procedure

    async def deactivate_procedure(self, procedure_id: str) -> bool:
        """Deactivate a procedure (soft delete)."""
        procedure = await self.get_procedure_by_id(procedure_id)
        if not procedure:
            return False

        procedure.is_active = False
        procedure.updated_at = datetime.utcnow()
        self.db.commit()
        return True

    async def get_procedure_requirements(self, procedure_id: str) -> Dict[str, Any]:
        """Get any special requirements for a procedure."""
        procedure = await self.get_procedure_by_id(procedure_id)
        if not procedure:
            return {}

        # Example requirements based on procedure category
        requirements = {
            "pre_operative": [],
            "post_operative": [],
            "equipment": [],
            "staff_requirements": 1,  # Default to 1 staff member
            "room_requirements": ["dental_chair"]
        }

        if procedure.category == ProcedureCategory.SURGICAL:
            requirements["pre_operative"].extend([
                "medical_clearance",
                "pre_operative_instructions"
            ])
            requirements["post_operative"].extend([
                "post_operative_instructions",
                "follow_up_appointment"
            ])
            requirements["equipment"].extend([
                "surgical_tray",
                "sutures"
            ])
            requirements["staff_requirements"] = 2  # Surgeon + assistant

        elif procedure.category == ProcedureCategory.ENDODONTIC:
            requirements["equipment"].extend([
                "endodontic_tray",
                "rubber_dam",
                "apex_locator"
            ])
            requirements["room_requirements"].append("xray_capable")

        return requirements

    async def get_procedure_statistics(self) -> Dict[str, Any]:
        """Get statistics about procedures."""
        total_procedures = self.db.query(Procedure).count()
        active_procedures = self.db.query(Procedure).filter(Procedure.is_active == True).count()
        
        category_counts = {}
        for category in ProcedureCategory:
            count = self.db.query(Procedure).filter(
                Procedure.category == category,
                Procedure.is_active == True
            ).count()
            category_counts[category.value] = count

        return {
            "total_procedures": total_procedures,
            "active_procedures": active_procedures,
            "category_counts": category_counts
        } 